lib/auth/oidc_google.go

```diff
@@ -0,0 +1,256 @@
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/coreos/go-oidc/jose"
+	"github.com/gravitational/trace"
+	"golang.org/x/oauth2/google"
+	directory "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/cloudidentity/v1"
+	"google.golang.org/api/option"
+)
+
+const (
+	// googleWorkspaceIssuerURL is the issuer URL for Google Workspace accounts.
+	googleWorkspaceIssuerURL = "https://accounts.google.com"
+
+	// googleGroupsClaim is the OIDC claim that we inject into the claims
+	// returned for Google Workspace users, containing the email addresses of
+	// the Google Groups that the user belongs to.
+	googleGroupsClaim = "groups"
+)
+
+// isGoogleWorkspaceConnector returns true if the connector is a OIDC connector
+// for Google Workspace, configured to fetch extra claims.
+func isGoogleWorkspaceConnector(connector types.OIDCConnector) bool {
+	// If google_service_account_uri and google_service_account are not set, we
+	// assume that this is a non-Google Workspace OIDC provider using the same
+	// issuer URL as Google Workspace (e.g.
+	// https://developers.google.com/identity/protocols/oauth2/openid-connect).
+	return connector.GetIssuerURL() == googleWorkspaceIssuerURL &&
+		(connector.GetGoogleServiceAccountURI() != "" || connector.GetGoogleServiceAccount() != "")
+}
+
+// addGoogleWorkspaceClaims will fetch extra data from proprietary Google APIs
+// and it will add claims based on the fetched data. The current implementation
+// adds a "groups" claim containing the Google Groups that the user is a member
+// of.
+func addGoogleWorkspaceClaims(ctx context.Context, connector types.OIDCConnector, claims jose.Claims) (jose.Claims, error) {
+	email, exists, err := claims.StringClaim("email")
+	if err != nil || !exists {
+		return nil, trace.BadParameter("no `email` in oauth claims for Google Workspace account")
+	}
+
+	var googleGroups []string
+	switch connector.GetVersion() {
+	// for the V3 connector we first check to see if the service account can use
+	// the Cloud Identity API (fetching direct and indirect groups for all
+	// domains) and if that's not the case we fall back to the Admin SDK
+	// Directory API (fetching direct groups for all domains)
+	case types.V3:
+		credentials, err := getGoogleWorkspaceCredentials(ctx, connector, cloudidentity.CloudIdentityGroupsReadonlyScope)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if credentials != nil {
+			log.Debugf("fetching transitive Google groups for %v", email)
+			googleGroups, err = groupsFromGoogleCloudIdentity(ctx, email, option.WithCredentials(credentials))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		} else {
+			credentials, err := getGoogleWorkspaceCredentials(ctx, connector, directory.AdminDirectoryGroupReadonlyScope)
```

Comment:
Similar logic to the V2 logic below, consider parameterizing into a function.

Reply:
I had tried refactoring it a bit but this is still the clearest way to write it that I've found, I'm afraid. 😢